### DISCUSSION TITLE
커스텀 예외 구조

### DISCUSSION THREAD

**Discussion by kong-hana01** [i-am-here-too/iht-backend#13] [개선사항 의견]
구현하고자하는 커스텀 예외 구조는 다음과 같습니다.

![image](https://github.com/haru-attendance/haru-backend/assets/79015120/0b9bb49c-cc2c-4561-8ed5-38268758a84e)
![image](https://github.com/haru-attendance/haru-backend/assets/79015120/75c55ae6-ea2e-438e-9ebe-b9e30efcdeb0)

`가장 상위에 HaruAttendanceException`을 두고, 이를 상속한 각 `서비스 계층`의 예외를 둡니다.
그리고 그 서비스 내부에서는 `각 inner class를 두어 각 상황에 맞는 커스텀 예외`를 미리 만들어두고 사용하고자 합니다.

### 장점
- 커스텀 예외 내부에서는 메세지와 httpStatus를 관리하게 하여 각 서비스 로직에서는 예외와 관련된 관심사를 분리할 수 있습니다.

### 단점
- 서비스 예외 간에 중복된 inner class 예외가 발생할 수 있습니다.
  - 예를 들어 MemberServiceException과 GroupServiceException 모두에 NonMemberException 가 필요할 수 있습니다.

일단 서비스 간에 의존성을 분리하려면 이 방법이 제일 좋아보이기는 하는데, 에코 의견은 어떤가요?

**Comment by echo724** [comment]
설 끝나고 이제서야 확인했습니다 ㅎㅎ 혹시 기다리셨다면 죄송합니다.
일단 서비스 별로 예외를 분리한 것은 좋습니다! 각 서비스 별로 예외 상황이 다르고 혹여나 나중에 분리를 할 경우에도 각 서비스 별로 다른 예외를 가지고 있어서 분리하기 쉬울 것 같습니다.
단점의 경우에는 제 생각에 단점으로 생각되지 않습니다! MemberService와 GroupService에서 비슷한 의미로 사용되는 예외라도 서비스 문맥에 따라 다를 수 있다고 생각해서 중복으로 있어도 괜찮을꺼 같아요! 예외 설계 좋습니다~ 이대로 진행하죠!

앗 다만 그룹 이름 바뀐 것도 반영 부탁 드려요~

**Reply by kong-hana01** [reply]
좋습니다! 단점에 대해 말씀해주신 부분도 다시 생각해보니 공감합니다 ㅎㅎ
사실 이 구조를 생각했던 게 `도메인 단위로 예외를 분리했을 때 서비스 별로 문맥에 맞게 예외를 발생시키기 어렵다`는 점 때문이었는데,
그 부분과 맞아떨어지네요!

그러면 해당 내용으로 그룹 이름도 수정해서 pr 올리겠습니다!
